### ISSUE TITLE
H2PushPriority interleaved might not actually work.

### ISSUE THREAD

**Issue by morlovich** [icing/mod_h2#82]
(This is testing with what's bundled with Apache 2.4.18; should I re-test with git version instead?)

Anyway, I am testing with a very simple setup: a giant HTML page, with two pushes initiated via Header Add (one CSS, one image), and with "H2PushPriority text/css  interleaved" in the configuration, since I want the stylesheet to load concurrently with the webpage.

Looking at the logs, mod_http2 processes the page, sends two PUSH_PROMISES, and schedules the GETs for the subresources --- then sends the entirety of the webpage. Only after it I see messages from h2_stream.c/h2_session.c about 200s on the pushes resources, and it notices that CSS ought to be interleaved --- which is now too late.

**Comment by icing** [comment]
mod_http2, as you have seen, identifies the PUSH_PROMISES and, simultaneously sends them to the client and schedules the processing internally. It then depends on the resources requested, the current server load and other things how fast this processing returns. 

Also, the speed of the connection plays a role. If it is fast, the bandwidth will be used for the main resource before the pushes are ready.

So, the interleaving is not a guarantee, just a way of saying that those pushes have the same priority as the main stream. But priority does not mean equal use of bandwidth, as that also requires data to be available.

It would be nice to have all data immediately available for streaming out, but processing needs time and uses resources.

**Comment by morlovich** [comment]
Hmm, it's possible that my test setup was kind of worst case for this ---
VM with fast connection but slow CPU share, but the HTML file was pretty
huge. I'll try some other setups.

**Comment by morlovich** [comment]
Re-tested with faster VM, slower connection (and 'before' push priority), and also don't seem to be seeing the pushed resource getting sent over until HTML is done.

**Comment by icing** [comment]
Ok, just re-verified my test cases and priorities work for me. Something is different for you. Which version of libnghttp2 do you have? Note, that ```H2PushPriority``` states "...a nghttp2 library version 1.5.0 or newer is necessary."

**Comment by morlovich** [comment]
nghttp version appears to be 1.5.0.

**Comment by icing** [comment]
Hmm, I realized that enabling the priority handling is a feature that needs to be selected at compile time. If your 2.4.18 was build with an older nghttp and you now have a new lib installed, it will not help.

So, coming back to your original question: yes, it would help if you could build the github version against the - then known nghttp2 1.5.0 or newer - and try that one.

I'll make an issue here that mod_http2 should INFO log not only the version, but also the capabilities compiled in. That would make such an analysis easy.

**Comment by morlovich** [comment]
On Wed, Mar 16, 2016 at 4:51 AM, Stefan Eissing <notifications@github.com>
wrote:

> Hmm, I realized that enabling the priority handling is a feature that
> needs to be selected at compile time. If your 2.4.18 was build with an
> older nghttp and you now have a new lib installed, it will not help.
>

Fairly certain it was built with 1.5.0 --- it was done on a clean VM, with
a build script that aims to build nghttp 1.5.

So, coming back to your original question: yes, it would help if you could
> build the github version against the - then known nghttp2 1.5.0 or newer -
> and try that one.
>
Still seems to happen, even with git build --- logs itself as mod_http2
(v1.2.8, nghttp2 1.5.0).

FWIW, here is my test setup:
URL is https://www.morlovich-work-test.com/amp2a.html (just using the amp
site as test data, as it's simple but not trivial)
Apache config has:
H2Push on
<Location /amp2a.html>
  Header Add Link "</cdn.ampproject.org/v0.js>; rel=preload; as=script"
</Location>
H2PushPriority application/javascript before

Tested both directly and with WebPageTest network shaping (which slows down
the HTML delivery considerably --- hmm, maybe that gets buffer-bloated,
though?).

> I'll make an issue here that mod_http2 should INFO log not only the
> version, but also the capabilities compiled in. That would make such an
> analysis easy.
>
That does sound like something that can save you and other people a lot of
potential headache.

**Comment by icing** [comment]
I had a recent "aha!" moment regarding stream output  and released changes in v1.8.8 that will also apply to the setup you described. Could you make a check with that release? Thanks!

**Comment by icing** [comment]
Feedback from others positive. no response here. closing.
